File: java/src/main/java/dev/nezo/burmaldaholic/core/bots/logic/Personality.java
package dev.nezo.burmaldaholic.core.bots.logic;

import java.util.Locale;

/**
 * Bot personality (BOTS.md §4.3, §4.7): changes style, never strength class (test BOTS.md §12.2).
 * Poker: threshold / aggression / bluff modifiers on top of the level. Roulette / craps / baccarat: the
 * betting style ({@code gui.burmaldaholic.bots.betstyle.<game>.<id>}). PvP: taunt flavour only.
 * {@code bots.personalities} = false → every bot is {@link #TAG}.
 *
 * @param weights     appearance weight at EASY / NORMAL / HARD
 * @param openDelta   Δ Chen open threshold (LAG −1.5: round toward the looser integer at use)
 * @param aggression  raise : call aggression factor
 * @param bluffFactor × bluff frequency
 * @param callWider   × calling range width (STATION 1.5)
 */
public enum Personality {
	ROCK(new int[] {25, 30, 0}, 2, 0.5, 0.3, 1.0),
	STATION(new int[] {35, 0, 0}, -3, 0.5, 0.2, 1.5),
	MANIAC(new int[] {40, 0, 0}, -3, 2.0, 3.0, 1.0),
	TAG(new int[] {0, 70, 60}, 0, 1.0, 1.0, 1.0),
	LAG(new int[] {0, 0, 40}, -1.5, 1.4, 1.5, 1.0);

	private final int[] weights;
	public final double openDelta;
	public final double aggression;
	public final double bluffFactor;
	public final double callWider;

	Personality(int[] weights, double openDelta, double aggression, double bluffFactor, double callWider) {
		this.weights = weights;
		this.openDelta = openDelta;
		this.aggression = aggression;
		this.bluffFactor = bluffFactor;
		this.callWider = callWider;
	}

	public String id() {
		return name().toLowerCase(Locale.ROOT);
	}

	/** {@code gui.burmaldaholic.bots.personality.<id>} (+ {@code .desc}). */
	public String translationKey() {
		return "gui.burmaldaholic.bots.personality." + id();
	}

	/** Betting style name at an atmosphere game ({@code roulette | craps | baccarat}). */
	public String betStyleKey(String game) {
		return "gui.burmaldaholic.bots.betstyle." + game + "." + id();
	}

	public int weight(BotDifficulty level) {
		return switch (level) {
			case EASY -> weights[0];
			case NORMAL, MIXED -> weights[1];
			case HARD -> weights[2];
		};
	}

	/** Draws a personality for a concrete level from the BOT rng ({@code enabled} = {@code bots.personalities}). */
	public static Personality pick(BotRng rng, BotDifficulty level, boolean enabled) {
		if (!enabled) {
			return TAG;
		}
		int total = 0;
		for (Personality p : values()) {
			total += p.weight(level);
		}
		if (total <= 0) {
			return TAG;
		}
		int roll = rng.nextInt(total);
		for (Personality p : values()) {
			roll -= p.weight(level);
			if (roll < 0) {
				return p;
			}
		}
		return TAG;
	}

	public static Personality byId(String id) {
		for (Personality p : values()) {
			if (p.id().equalsIgnoreCase(id)) {
				return p;
			}
		}
		return TAG;
	}
}

File: bedrock/src/core/logic/bots/personality.ts
/** Bot personalities (BOTS.md §4.3, §4.7). PURE. Never crosses levels (test BOTS.md §12.2). */
import type { Rng } from '../rng';
import type { BotLevel, Personality } from './types';

export interface PersonalityDef {
  /** appearance weight at EASY / NORMAL / HARD */
  readonly weights: readonly [number, number, number];
  /** Δ Chen open threshold (LAG −1.5: round toward the looser integer at use) */
  readonly openDelta: number;
  /** raise : call aggression factor */
  readonly aggression: number;
  readonly bluffFactor: number;
  /** × calling range width */
  readonly callWider: number;
}

export const PERSONALITY_DEFS: Readonly<Record<Personality, PersonalityDef>> = {
  ROCK: { weights: [25, 30, 0], openDelta: 2, aggression: 0.5, bluffFactor: 0.3, callWider: 1 },
  STATION: { weights: [35, 0, 0], openDelta: -3, aggression: 0.5, bluffFactor: 0.2, callWider: 1.5 },
  MANIAC: { weights: [40, 0, 0], openDelta: -3, aggression: 2, bluffFactor: 3, callWider: 1 },
  TAG: { weights: [0, 70, 60], openDelta: 0, aggression: 1, bluffFactor: 1, callWider: 1 },
  LAG: { weights: [0, 0, 40], openDelta: -1.5, aggression: 1.4, bluffFactor: 1.5, callWider: 1 },
};

const LEVEL_INDEX: Record<BotLevel, 0 | 1 | 2> = { EASY: 0, NORMAL: 1, HARD: 2 };

/** Draw a personality for a level from the BOT rng (`enabled` = `bots.personalities`; off → TAG). */
export function pickPersonality(rng: Rng, level: BotLevel, enabled = true): Personality {
  if (!enabled) return 'TAG';
  const entries = (Object.keys(PERSONALITY_DEFS) as Personality[]).map((p) => [p, PERSONALITY_DEFS[p].weights[LEVEL_INDEX[level]]] as const);
  const total = entries.reduce((s, [, w]) => s + w, 0);
  if (total <= 0) return 'TAG';
  let r = rng.next() * total;
  for (const [p, w] of entries) {
    r -= w;
    if (r < 0) return p;
  }
  return 'TAG';
}

/** Betting style name at an atmosphere game (BOTS.md §4.7). */
export const betStyleKey = (game: 'roulette' | 'craps' | 'baccarat', p: Personality): string => `gui.burmaldaholic.bots.betstyle.${game}.${p.toLowerCase()}`;
